godtools/App/Share/Domain/UseCases/GetUserDetailsUseCase/GetUserDetailsUseCase.swift

```diff
@@ -0,0 +1,79 @@
+//
+//  GetUserDetailsUserCase.swift
+//  godtools
+//
+//  Created by Rachael Skeath on 11/21/22.
+//  Copyright © 2022 Cru. All rights reserved.
+//
+
+import Foundation
+import Combine
+
+class GetUserDetailsUseCase {
+    
+    private let repository: UserDetailsRepository
+    private let localizationServices: LocalizationServices
+    
+    init(repository: UserDetailsRepository, localizationServices: LocalizationServices) {
+        
+        self.repository = repository
+        self.localizationServices = localizationServices
+    }
+    
+    
+    func getUserDetailsPublisher() -> AnyPublisher<UserDetailsDomainModel?, Never> {
```

Comment:
Hey @rachaelblue a couple suggestions here.

*  What if instead of adding the ```fetchRemoteUserDetailsPublisher``` here, we provide a ```reloadFromRemote``` flag on the ```UserDetailsRepository getUserDetailsChanged()``` which would then trigger the changed publisher again after remote has finished.  Then we get some re-usability if we added this to any other UseCases.
```swift
func getUserDetailsChanged(reloadFromRemote: Bool) -> AnyPublisher<Void, Never> {
        
        if reloadFromRemote {
            
            fetchRemoteUserDetails()
                .sink { finished in
                    
                } receiveValue: { (userDataModel: UserDetailsDataModel) in
                    
                }
                .store(in: &cancellables)

        }
        
        return cache.getUserDetailsChanged()
    }
```
* Another is changing the return type to a non-optional value ```UserDetailsDomainModel``` and instead provide a default value the UI can display.  This way the ViewModel doesn't have to do any unwrapping and can always have a value ready to display.  Which in our case is just an empty string.

```swift
private func getDefaultUserDetails() -> UserDetailsDomainModel {
        return UserDetailsDomainModel(
            joinedOnString: ""
        )
    }
    
    func getUserDetailsPublisher() -> AnyPublisher<UserDetailsDomainModel, Never> {
        
        return repository.getUserDetailsChanged(reloadFromRemote: true)
            .flatMap { () in
                    
                guard let userDetails = self.repository.getAuthUserDetails() else {
                    
                    return Just(self.getDefaultUserDetails())
                        .eraseToAnyPublisher()
                }
                
                let joinedOnString = self.buildJoinedOnString(from: userDetails)
                
                return Just(UserDetailsDomainModel(joinedOnString: joinedOnString))
                    .eraseToAnyPublisher()
            }
            .eraseToAnyPublisher()
    }
```

Reply:
Good idea about the `UseretailsDomainModel`, I went ahead and made that change!

It feels a little odd to me for `getUserDetailsChanged` to have such a huge side effect as pulling from the remote.  Also then the repository will be responsible for holding onto that cancellable instead of the class that initiated it.  It seems like if there's a scenario where you'd want to pull from the remote and get the userDetails changed publisher, then it'd probably make sense to just use `GetUserDetailsUseCase`?

Reply:
Hey @rachaelblue I need to think about this one some more.  I feel like the data layer should be responsible for invalidating the cache which would trigger a new fetch from remote.  Currently we don't have a common solution for this. It would be nice if there could be some common cache control policy on the data layer that allows more control of how data can be fetched whether it be from cache only, remote only, cache and remote, etc.  I was looking through the Apollo GraphQL SDK a little bit and I like their setup https://www.apollographql.com/docs/ios/fetching/fetching-data/#cache-policies.   For now I'm fine with leaving the fetch in the UseCase until we can come up with a better solution too. 

Another thing I did just think of is the location of this UseCase in ```Share/Domain/``` and possibly the naming.  I would like us to start getting more specific with UseCases and having those live in a ```{Feature_Name}/Domain/UseCases/``` folder . 

The Account presentation files currently live in ```Features/Menu/``` so a good place for this UseCase would be in ```Features/Menu/Domain/UseCases/```.  As the Account grows feel free to create a specific Feature for the Account as well if you'd like because I know you will be adding more logic to the Account section soon.

As far as the UseCase naming, we could have 2 use cases, 1 responsible for the account creation date and another for the account user name ```GetUserAccountCreationDateUseCase``` and ```GetUserAccountNameUseCase``` or a single UseCase ```GetUserAccountDetailsUseCase``` for handling fetching the name and creation date and applying business rules.  This way we get away from the "Quazi" clean pattern.

Reply:
One thing that Daniel B has noted on the web side is that Apollo is a fairly large library there. It might be something to watch for here as well for app size.

Reply:
Hey @wrandall22 that's good to know about the Apollo library size.  I'm still hesitant about adding Apollo into our own projects. There would be some benefits, but as far as cache syncing we'd still have to find alternative solutions for that.

Reply:
Hey @rachaelblue yes a ticket would be good to have down for this.  Our repositories will most likely be duplicating logic around fetching and syncing of data, so some type of common solution for this would be nice to have that our repositories can make use of.

Reply:
Ticket here: https://jira.cru.org/browse/GT-1887